Set controllerRef in RCs owned by DC

tnozicka opened the pull request:
TODO:
 - [ ] Add unit tests
 - [ ] Finalize https://github.com/kubernetes/kubernetes/pull/45894

# Goals

* The main goal of ControllerRef (controller reference) is to solve the problem
  of controllers that fight over controlled objects due to overlapping selectors.
  Fighting controllers can [destabilize the apiserver](https://github.com/kubernetes/kubernetes/issues/24433),
  [thrash objects back-and-forth](https://github.com/kubernetes/kubernetes/issues/24152),
  or [cause controller operations to hang](https://github.com/kubernetes/kubernetes/issues/8598).

* A secondary goal of ControllerRef is to provide back-links from a given object
  to the controller that manages it, which can be used for:
  * Efficient object->controller lookup, without having to list all controllers.
  * Generic object grouping (e.g. in a UI), without having to know about all
    third-party controller types in advance.

* By being a subtype of OwnerReference it fixes a bug when the underlying RC are not removed when the object is deleted using API.

CC @mfojtik @kargakis

kargakis commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -631,7 +631,7 @@ func GetOpenshiftBootstrapClusterRoles() []authorizationapi.ClusterRole {
 			},
 			Rules: []authorizationapi.PolicyRule{
-				authorizationapi.NewRule("get", "list", "watch", "update").Groups(kapiGroup).Resources("replicationcontrollers").RuleOrDie(),
+				authorizationapi.NewRule("get", "list", "watch", "update", "delete").Groups(kapiGroup).Resources("replicationcontrollers").RuleOrDie(),
```
Why does the deployer need to delete RCs now?

tnozicka replied:
@kargakis you can't set controller ref on a resource you aren't allowed to delete

kargakis replied:
Why is the deployer setting owner references?

tnozicka replied:
doesn't it create RCs?

kargakis replied:
No, the RCs are created by the DC controller.

tnozicka replied:
you are right; I forgot which SA am I talking about. This one needs it because otherwise I get
```
oc logs po/busyapp-1-deploy
--> Scaling busyapp-1 to 5
error: couldn't scale busyapp-1 to 5: Scaling the resource failed with: replicationcontrollers "busyapp-1" is forbidden: cannot set an ownerRef on a resource you can't delete: User "system:serviceaccount:test:deployer" cannot delete replicationcontrollers in project "test", <nil>; Current resource version 848
```
As it is not creating the resource and RC already has ownerRef set it might be because it is trying to remove it. This might happen if it forgets to do a deepcopy while updating since this is a pointer field. I will investigate it further

tnozicka replied:
@kargakis what I've found out so far is that `openshift-deploy` tries to do a PUT on RC without the ownerReferences and get 403

mfojtik replied:
https://github.com/openshift/origin/blob/master/pkg/cmd/server/bootstrappolicy/controller_policy.go#L67 seems to be fixed in HEAD

tnozicka replied:
@mfojtik actually I think that's a different one and that one needs the "delete" as well

kargakis replied:
```
error: couldn't scale busyapp-1 to 5: Scaling the resource failed with: replicationcontrollers "busyapp-1" is forbidden: cannot set an ownerRef on a resource you can't delete: User "system:serviceaccount:test:deployer" cannot delete replicationcontrollers in project "test", <nil>; Current resource version 848
```
say whaaat?

kargakis commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -64,7 +64,7 @@ func init() {
 		Rules: []rbac.PolicyRule{
 			rbac.NewRule("create", "get", "list", "watch", "patch", "delete").Groups(kapiGroup).Resources("pods").RuleOrDie(),
-			rbac.NewRule("get", "list", "watch", "update").Groups(kapiGroup).Resources("replicationcontrollers").RuleOrDie(),
+			rbac.NewRule("get", "list", "watch", "update", "delete").Groups(kapiGroup).Resources("replicationcontrollers").RuleOrDie(),
```
Not sure that you need this permission. The owner references should be set by the DC controller.

mfojtik replied:
this is fine to investigate later, please create issue (most likely you need delete to create ownerRef)

mfojtik commented:
@tnozicka @kargakis we need to ship this today or we will have to wait for DCUT to pass.

mfojtik commented:
[test]

tnozicka commented:
[test]

tnozicka commented:
[test]

tnozicka commented:
@mfojtik @kargakis any other comments except "delete" in RBAC which I will address in a separate PR?

tnozicka commented:
@kargakis @mfojtik I believe I've found the issue with `openshift-deploy`. After I've copied the newly build binary from my machine into the image and it worked. As also does the `openshift/origin-deployer:latest` image. The issue is that we don't reference latest, but `v3.6.0-alpha.1` which is 6 weeks old. Running with this image causes the problem because it was compiled before the kube rebase was updated to know about ownerReferences (https://github.com/openshift/origin/blame/master/vendor/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L258). This would mean that it is compiled without knowing about ownerReferences and when it does an update the issue manifest itself:
 - it decodes RC from wire
 - skips fields it doesn't know about (controllerRef) while deserializing
 - updates what it needs on that struct (to scale)
 - serializes the object with those changes
 - sends PUT for that RC!
And the controllerRef is lost! Because it is trying to modify the controllerRef in the PUT (from a valid record to nil - that means indirect delete by garbage collector) it is rightfully denied such PUT with 403.
Using PATCH would also mitigate the issue.

(It seemed to work with elevated permissions because DC controller created the controllerRef, than Deployer deleted it and DC controller adopted it because it didn't have any controllerRef and matched the selector.)

kargakis commented:
@tnozicka nice. This means that if we don't provide the permissions in the deployer sa, we are going to break older images :(

kargakis commented:
This lacks an extended test and you also need to set finalizers like in https://github.com/kubernetes/kubernetes/pull/45764

kargakis commented:
s/extended/integration/

mfojtik commented:
@kargakis i'm fine with adding more permissions there with reasonable explanation/todo to remove them after all users are migrated. Having more escalated privileges is still better than having no privileges at all.

I would defer all tests as follow ups as we need to unblock the UI team and web console. I think this PR will qualify as blocker bug fix.

tnozicka commented:
Ok, I have looked into it and it seems like the deployer image version and name come from master config. That kind of complicates things because we might not get rid those old images any time soon.

But the real issues seems to be that the deployer images are not backwards compatible (or forwards - depends how you look at it). Meaning: if you do a forwards compatible API change they will delete the new data and act as nothing happened. This is a data loss. In this case it gets filled back when adopted back by DC but it might be more serious in other cases.

The reason is the PUT call. Whenever you do GET+PUT on a resource (and you don't want to completely overwrite it) you break compatibility! You loose any fields you don't know about. Especially with deployer images which are decoupled from the master binary and you run different versions. But this is also applicable elsewhere - if you do a rollback of infra and use PUT on such resources you loose the data that has been added by the newer version.

My take: don't use PUT, but PATCH.

As we are already in that situation, here is a proposed path forward:
1. Leave the elevated privileges here with TODO and a comment referencing this PR for an explanation. The issue with deleting ownerReferences with every rollout will still be there but that deleted ref gets repared by adoption literally immediately.
1. Add finalizers (https://github.com/kubernetes/kubernetes/pull/45764/files)
1. Get it merged. It fixes a bug when RCs are not deleted when DC is and it unblocks UI team.
1. Send a PR moving deployer to use PATCH (+ add patch to RBAC)
1. Create a card to remove unneeded privileges in a few releases

@mfojtik @kargakis WDYT

kargakis commented:
If all the deployer does is scaling the RC then it should use its scale endpoint. The fact that we are adding priviledges because we need to keep old images working worries me but I don't see any other solution now. cc: @smarterclayton

mfojtik commented:
@tnozicka agree, lets move this forward, don't have to be perfect but we need to unblock UI team and have the owner refs. We can fiddle the permissions and tests later.

tnozicka commented:
I added the comments and I've tried to setup the finalizer but I've got:
```
metadata.finalizers[0]: Invalid value: "foregroundDeletion": name is neither a standard finalizer name nor is it fully qualified
```
Tried prefixing it with kubernetes.io/ it just doesn't do anything. We migh be missing that finalizer in origin but I wasn't able to confirm it yet.

tnozicka commented:
Ok I got to the bottom of this - seems like upstream is broken https://github.com/kubernetes/kubernetes/blob/58167fcfa10551cc092ee7484e8b41082d0ac223/pkg/api/helper/helpers.go#L224

They are missing that finalizer in standardFinalizers. It works perfectly if add `metav1.FinalizerDeleteDependents` right after it.

 I will send an PR and see.

tnozicka commented:
@mfojtik @kargakis should we have admissionControl to add `metav1.FinalizerDeleteDependents` to every DC so the objects are deleted in correct order? Or we don't care?

Right now it deletes DC, then PODs (that may take long time - default is 30 s) and then RCs. If in the interim user does `oc get all` he sees just PODs terminating and RCs waiting for it; no DC.

tnozicka commented:
Flake https://github.com/openshift/origin/issues/14367

tnozicka commented:
[test]

tnozicka commented:
flake https://github.com/openshift/origin/issues/13980
[test]

mfojtik commented on pkg/cmd/server/bootstrappolicy/controller_policy.go:
```diff
@@ -64,7 +64,11 @@ func init() {
+			// "delete" is required here for compatibility with older deployer images
+			// (see https://github.com/openshift/origin/pull/14322#issuecomment-303968976)
+			// TODO: remove "delete" few releases after 3.6
+			rbac.NewRule("get", "list", "watch", "update", "delete").Groups(kapiGroup).Resources("replicationcontrollers").RuleOrDie(),
```
nit: I would make this an extra rule just with delete

rbac.NewRule("delete").Group(kapi).Resources("replicationcontrollers")

mfojtik commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -631,7 +631,10 @@ func GetOpenshiftBootstrapClusterRoles() []authorizationapi.ClusterRole {
+				// "delete" is required here for compatibility with older deployer images
+				// (see https://github.com/openshift/origin/pull/14322#issuecomment-303968976)
+				// TODO: remove "delete" few releases after 3.6
+				authorizationapi.NewRule("get", "list", "watch", "update", "delete").Groups(kapiGroup).Resources("replicationcontrollers").RuleOrDie(),
```
same here, i would separate this from the rest of rbac

mfojtik commented on pkg/controller/controller_ref_manager.go:
```diff
@@ -0,0 +1,168 @@
+
+// RealRCControl is the default implementation of RCControlInterface.
+type RealRCControl struct {
+	KubeClient kclientset.Interface
```
do we need this exported?

tnozicka replied:
We do; otherwise you couldn't initialize `RealRCControl` in `factory.go`

mfojtik replied:
@tnozicka we can create initializer func()

tnozicka replied:
we could, although I don't think that it would bring any value here. Also I would rather leave the upstream code mostly unchanged

mfojtik replied:
ok, I think for many of our controllers we have initializers, that is why I asked for one :-) But if this is copying the upstream pattern i'm ok with it.

mfojtik commented on pkg/controller/controller_ref_manager.go:
```diff
@@ -0,0 +1,168 @@
+// To make sure RealRCControl implements RCControlInterface
+var _ RCControlInterface = &RealRCControl{}
+
+func (r RealRCControl) PatchReplicationController(namespace, name string, data []byte) error {
```
add godoc

mfojtik commented on pkg/controller/controller_ref_manager.go:
```diff
@@ -0,0 +1,168 @@
+// the patching fails.
+func (m *RCControllerRefManager) AdoptReplicationController(rs *kapi.ReplicationController) error {
+	if err := m.CanAdopt(); err != nil {
+		return fmt.Errorf("can't adopt ReplicationController %v/%v (%v): %v", rs.Namespace, rs.Name, rs.UID, err)
```
use %s/%s (%s) for ns/name and uuid.

mfojtik commented on pkg/controller/controller_ref_manager.go:
```diff
@@ -0,0 +1,168 @@
+// ReleaseReplicationController sends a patch to free the ReplicationController from the control of the Deployment controller.
+// It returns the error if the patching fails. 404 and 422 errors are ignored.
+func (m *RCControllerRefManager) ReleaseReplicationController(rc *kapi.ReplicationController) error {
+	glog.V(2).Infof("patching ReplicationController %s_%s to remove its controllerRef to %s/%s:%s",
```
%s_%s is weird for ns/name

also this should be V(4)

mfojtik commented on pkg/controller/controller_ref_manager.go:
```diff
@@ -0,0 +1,168 @@
+			// If the ReplicationController no longer exists, ignore it.
+			return nil
+		}
+		if kerrors.IsInvalid(err) {
```
merge this into one if or make a switch

tnozicka replied:
@mfojtik essentially this whole file is from upstream with a bit of `sed s/RS/RC/g`. Do we want to deviate here w.r.t. future diffs?

mfojtik replied:
this is fine then

mfojtik commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -84,11 +89,29 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+	// They will be orphaned by ClaimReplicationControllers().
+	rcList, err := c.rcLister.ReplicationControllers(config.Namespace).List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("DeploymentConfigController can't list all RCs: %s", err)
```
%v instead of %s

mfojtik replied:
also the format we use is:

`"error while deploymentConfigController listing the replication controllers: %v"`

mfojtik commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -84,11 +89,29 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+			return nil, err
+		}
+		if fresh.UID != config.UID {
+			return nil, fmt.Errorf("original DeploymentConfig %v/%v is gone: got uid %v, wanted %v", config.Namespace, config.Name, fresh.UID, config.UID)
```
%s/%s and uuid's are %s

also `error matching the deployment config foo/bar: current uuid UUID does not match UUID`... (if this is matching the upstream error, ignore this comment ;-)

tnozicka replied:
yes, upstream

mfojtik commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -84,11 +89,29 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+	existingDeployments, err := cm.ClaimReplicationControllers(rcList)
 	if err != nil {
-		return err
+		return fmt.Errorf("DeploymentConfigController failed to claim RCs: %s", err)
```
%v (we use %v for errors in general, and %s for strings (sometimes %q for string but I forgot what the rule was ;-))

mfojtik commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -125,6 +148,15 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+				// We need to make sure we own that RC or adopt it if possible
+				isOurs, err := cm.ClaimReplicationController(rc)
+				if err != nil {
+					return fmt.Errorf("DeploymentConfigController failed to claim the RC: %s", err)
```
nit: `error while deploymentConfigController claiming the RC %s/%s: %v` (add the rc name)

mfojtik commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -239,6 +271,18 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments []
+				// We need to make sure we own that RC or adopt it if possible
+				isOurs, err := cm.ClaimReplicationController(rc)
+				if err != nil {
+					return fmt.Errorf("DeploymentConfigController failed to claim the RC: %s", err)
```
same as above.

mfojtik commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -239,6 +271,18 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments []
+					return fmt.Errorf("DeploymentConfigController failed to claim the RC: %s", err)
+				}
+				if !isOurs {
+					return fmt.Errorf("DC %v/%v (%v) no longer owns RC %v/%v (%v)",
```
s/DC/DeploymentConfig/ also use %s

mfojtik commented on pkg/deploy/controller/deploymentconfig/factory.go:
```diff
@@ -51,6 +52,10 @@ func NewDeploymentConfigController(
 		rcLister:       rcInformer.Lister(),
 		rcListerSynced: rcInformer.Informer().HasSynced,
+		rcControl: oscontroller.RealRCControl{
+			KubeClient: internalKubeClientset,
```
i would create `NewRealRCControl()` and pass the client and recorded and make them unexported unless.

tnozicka replied:
upstream pattern as per https://github.com/openshift/origin/pull/14322#discussion_r118942767

mfojtik commented on pkg/deploy/util/util.go:
```diff
@@ -18,10 +18,16 @@ import (
 
+var (
+	// controllerKind contains the schema.GroupVersionKind for this controller type.
+	ControllerKind = osapiv1.SchemeGroupVersion.WithKind("DeploymentConfig")
```
@kargakis will this break with legacy kind? (no api group?)

Also do we need this exported? If yes, fix the godoc. Also I would rename this to `ownerRefControllerKind` or something to be less generic. 

maybe we don't need this at all if we use this only in one place...

tnozicka replied:
Has to be exported because `util` and `deploymentconfig` are different packages.

It is used in 3 places:
- https://github.com/tnozicka/origin/blob/set-owner-refs-in-rcs-owned-by-dc/pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go#L111
- https://github.com/tnozicka/origin/blob/set-owner-refs-in-rcs-owned-by-dc/pkg/deploy/util/util.go#L241-L242

and the naming aims to be consistent with [upstream](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L64). I can change it if you wish so?

mfojtik commented on pkg/deploy/util/util.go:
```diff
@@ -228,6 +234,19 @@ func EncodeDeploymentConfig(config *deployapi.DeploymentConfig, codec runtime.Co
 }
 
+func NewControllerRef(config *deployapi.DeploymentConfig) *metav1.OwnerReference {
+	blockOwnerDeletion := true
```
@kargakis i wish upstream has helper `IntPtr(true)` :-)

tnozicka replied:
@mfojtik I wish Go had a syntax for it ;) 

But these few lines are also from upstream.

mfojtik replied:
@tnozicka I saw BoolPtr() in some vendor package (not kube). Since we are doing this in a lot of places, it sounds reasonable to me:

```
func BoolPtr(r bool) *bool {
  return &r
}

func BoolTruePtr() *bool {
  return BoolPtr(true)
}
```

mfojtik commented on pkg/deploy/util/util.go:
```diff
@@ -228,6 +234,19 @@ func EncodeDeploymentConfig(config *deployapi.DeploymentConfig, codec runtime.Co
+	blockOwnerDeletion := true
+	isController := true
+	return &metav1.OwnerReference{
+		APIVersion:         ControllerKind.Version,
```
@sttts @ncdc will this break for API groups? i guess we store everything as legacy group, should the controller ref point to legacy or API group version?

sttts replied:
My guess is that it does not matter, as long as the referenced object exists. But I am not deep in the GC algorithm either.

mfojtik replied:
@sttts i guess that if we use non-versioned client to get the DC in GC it should work fine

sttts replied:
it uses the dynamic client.

mfojtik commented on pkg/deploy/util/util.go:
```diff
@@ -292,7 +311,8 @@ func MakeDeployment(config *deployapi.DeploymentConfig, codec runtime.Codec) (*a
 			},
-			Labels: controllerLabels,
+			Labels:          controllerLabels,
+			OwnerReferences: []metav1.OwnerReference{*NewControllerRef(config)},
```
i will extract this into variable, * on function seems weird ;-)

tnozicka replied:
I don't think there is anything wrong with dereferencing a function result, but I will extract it into variable.

mfojtik commented:
couple nits and code style but looks good over all. i would like to make sure this won't break for "old" DC with no API group before merging.

would create follow up issue for adding tests to unblock @spadgett

tnozicka commented:
@mfojtik I have addressed your comments

Open questions:
 - [order of deletion](https://github.com/openshift/origin/pull/14322#issuecomment-304247136)
 - [GC client](https://github.com/openshift/origin/pull/14322#pullrequestreview-40741453)

tnozicka commented:
[test]

tnozicka commented:
flake https://github.com/openshift/origin/issues/14397
[test]

tnozicka commented:
flake https://github.com/openshift/origin/issues/14403
[test]

mfojtik commented:
LGTM (this is part of P1 https://github.com/openshift/origin/issues/14133 and blocks the console team)

mfojtik commented:
@tnozicka can you squash pls?

tnozicka commented:
@mfojtik good to go?

For the record: GC client seems to work fine as tried with upgrade from 1.4; order of deletion will be decided on separately.

mfojtik commented:
@tnozicka yeah, please make sure you have follow up card for the tests for this... I think this is good base and we can build on this.

[merge][severity: bug]

- this unblocks the UI team
- https://github.com/openshift/origin/issues/14133 (P1, 3.6)

mfojtik commented:
[severity:blocker]

enj commented:
[merge][severity: blocker]

See previous comments.

mfojtik commented:
[merge] again, this time with proper tags

mfojtik commented:
[merge] against (seems like tagging screwup)

tnozicka commented:
Flake https://github.com/openshift/origin/issues/14122

tnozicka commented:
Flake https://github.com/openshift/origin/issues/14418

[merge] again

tnozicka commented:
[merge] again

mfojtik commented:
[merge]

kargakis commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -125,6 +148,15 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 					return err
 				}
+				// We need to make sure we own that RC or adopt it if possible
+				isOurs, err := cm.ClaimReplicationController(rc)
```
You have already claimed those RCs above, no?

tnozicka replied:
@kargakis the one we get with the `Get` could be a different one. What if it no longer matches the selector and was/should be released or it could have been changed, deleted/+created and we don't own it

kargakis replied:
Not sure about having the RC diverge just a couple of Go code after the initial adoption and this is a cache lookup but maybe this doesn't hurt as a last line of defense? What happens if the cache copy is not yet updated with the owner reference you set above? Will the 2nd patch simply succeed because there is no diff with the server copy or is there an actual error returned? 

Now I am thinking that we never supported adopting RCs for DCs, before owner refs, so I am wondering if we are going to break users that have fiddled with RC labels. Maybe warn in the release notes since we are departing from the old way of selecting RCs.

kargakis replied:
> Now I am thinking that we never supported adopting RCs for DCs, before owner refs, so I am wondering if we are going to break users that have fiddled with RC labels. Maybe warn in the release notes since we are departing from the old way of selecting RCs.

Nevermind this one, rechecked and you merely add owner refs on top - should be fine.

tnozicka replied:
@kargakis 

> What happens if the cache copy is not yet updated with the owner reference you set above?

This is in `RetryOnConflict` block so in the worst case caches should propagate while doing the backoff, right? We could use an uncached client here.

> What happens if the cache copy is not yet updated with the owner reference you set above? 

I think that the second patch should result in noop (and succeed).

> Not sure about having the RC diverge just a couple of Go code after the initial adoption and this is a cache lookup but maybe this doesn't hurt as a last line of defense?

Well not doing this and patching a RC we don't own due to race condition would break 2. or 3. [controller rule](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#the-three-laws-of-controllers)

But that also gets me thinking that we can break it under some conditions. There is still race condition in the patches we use (taken from upstream). I think they should really use optimistic concurrency in form of resourceVersion to be completely correct. Hope that works with `PATCH`. I'll take the discussion upstream to figure it out because the do the same.

kargakis commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -239,6 +287,18 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments []
 					return err
 				}
+				// We need to make sure we own that RC or adopt it if possible
+				isOurs, err := cm.ClaimReplicationController(rc)
```
Same comment as above

tnozicka replied:
same as well :)

kargakis commented on pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -177,10 +209,26 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+				// If the deployment was already created, just move on. The cache could be
+				// stale, or another process could have already handled this update.
+				return c.updateStatus(config, existingDeployments)
+			} else {
```
No need for else

tnozicka replied:
I know, I have changed it 2 times while writing it but somehow I wanted to explicitly state that this is for the case where it isn't ours :) and to be resilient to future changes
